### ISSUE TITLE
【BUG】where date in (int) will be cast to double type

### ISSUE THREAD

**Issue by xinghuayu007** [apache/doris#4805]
**Describe the bug**
`select * from table date in (20201030)`, `date` is type of DATE or DATETIME.　`date` and `20201030` will be casted to `DOUBLE` type. When `date` is partition key, it will cause exeption. It is better to cast 20201030 to date type.

**To Reproduce**
1. create a table 
`
CREATE TABLE `invalid_date` (
  `date` datetime NULL COMMENT "",
  `day` date NULL COMMENT "",
  `site_id` int(11) NULL COMMENT ""
) ENGINE=OLAP
DUPLICATE KEY(`date`, `day`, `site_id`)
COMMENT "OLAP"
PARTITION BY RANGE(`day`)
(PARTITION p20201030 VALUES [('2020-10-30'), ('2020-10-31')))
DISTRIBUTED BY HASH(`site_id`) BUCKETS 10
PROPERTIES (
"replication_num" = "1",
"in_memory" = "false",
"storage_format" = "V2"
);
`
2. execute a query:
`select date from invalid_date where day in (20201030);`
3. See error
![2020-10-27 15-39-32屏幕截图](https://user-images.githubusercontent.com/12771191/97270809-b5c33000-186a-11eb-8dca-95c664bbb40a.png)
![2020-10-27 15-39-57屏幕截图](https://user-images.githubusercontent.com/12771191/97270821-b8258a00-186a-11eb-992b-04db57727a39.png)

**Expected behavior**

**Screenshots**

**Desktop (please complete the following information):**

**Smartphone (please complete the following information):**

**Additional context**

**Comment by Yao-MR** [comment]
why must cast the int to date type?

in my idea, i think this the type cast problem, so the problem is what is the type should the cast the int to,  

so just cast the int to the type of the partition key type,

and i also doubt that why scan the all partion. 


myabe this is the right flow ,

1, we should know that the condition key has the partion key, and get the info from the catalog

**Comment by xinghuayu007** [comment]
Yes, this is the problem of type-casting. In current system, sql like `select * from table where date in (20201030);` date is type of DATE; date and 20201030 will be casted to `DOUBLE` type. That is not corret. I think the invalid type should be found in `Analyze Phase`, not be found in phase of 'Pruning Partitions'. This problem will not cause to scan all partitions.

**Comment by xinghuayu007** [comment]
> why must cast the int to date type?
> 
> in my idea, i think this the type cast problem, so the problem is what is the type should the cast the int to,
> 
> so just cast the int to the type of the partition key type,
> 
> and i also doubt that why scan the all partion.
> 
> myabe this is the right flow ,
> 
> 1, we should know that the condition key has the partion key, and get the info from the catalog

Yes, this is the problem of type-casting. In current system, sql like select * from table where date in (20201030); date is type of DATE; date and 20201030 will be casted to DOUBLE type. That is not corret. I think the invalid type should be found in Analyze Phase, not be found in phase of 'Pruning Partitions'. This problem will not cause to scan all partitions.
